packages/yew/src/html/component/marker.rs

```diff
@@ -0,0 +1,147 @@
+/// The type of the generic parameter is not required to be the same component that was in the other
+/// implementation. However, this behaviour may change in the future if more debug assertions were
+/// to be introduced. It is recommended that the generic parameter represents the component in the
+/// other implementation.
+///
+/// # Example
+///
+/// ```
+/// use yew::prelude::*;
+/// # use yew::html::ChildrenProps;
+/// #
+/// # #[function_component]
+/// # fn Comp(props: &ChildrenProps) -> Html {
+/// #     Html::default()
+/// # }
+/// #
+/// # #[function_component]
+/// # fn Provider(props: &ChildrenProps) -> Html {
+/// #     let children = props.children.clone();
+/// #
+/// #     html! { <>{children}</> }
+/// # }
+/// # type Provider1 = Provider;
+/// # type Provider2 = Provider;
+/// # type Provider3 = Provider;
+/// # type Provider4 = Provider;
+///
+/// #[function_component]
+/// fn ServerApp() -> Html {
+///     // The Server Side Rendering Application has 3 Providers.
+///     html! {
+///         <Provider1>
+///             <Provider2>
+///                 <Provider3>
+///                     <Comp />
+///                 </Provider3>
+///             </Provider2>
+///         </Provider1>
+///     }
+/// }
+///
+/// #[function_component]
+/// fn App() -> Html {
+///     // The Client Side Rendering Application has 4 Providers.
+///     html! {
+///         <Provider1>
+///             <Provider2>
+///                 <Provider3>
+///
+///                     // This provider does not exist on the server-side
+///                     // Hydration will fail due to Virtual DOM layout mismatch.
+///                     <Provider4>
+///                         <Comp />
+///                     </Provider4>
+///
+///                 </Provider3>
+///             </Provider2>
+///         </Provider1>
+///     }
+/// }
+/// ```
+///
+/// To mitigate this, we can use a `PhantomComponent`:
+///
+/// ```
+/// use yew::prelude::*;
+/// # use yew::html::{PhantomComponent, ChildrenProps};
+/// #
+/// # #[function_component]
+/// # fn Comp(props: &ChildrenProps) -> Html {
+/// #     Html::default()
+/// # }
+/// #
+/// # #[function_component]
+/// # fn Provider(props: &ChildrenProps) -> Html {
+/// #     let children = props.children.clone();
+/// #
+/// #     html! { <>{children}</> }
+/// # }
+/// # type Provider1 = Provider;
+/// # type Provider2 = Provider;
+/// # type Provider3 = Provider;
+/// # type Provider4 = Provider;
+///
+/// #[function_component]
+/// fn ServerApp() -> Html {
+///     html! {
+///         <Provider1>
+///             <Provider2>
+///                 <Provider3>
+///                     // We add a PhantomComponent for Provider4,
+///                     // it acts if a Provider4 component presents in this position.
+///                     <PhantomComponent<Provider4>>
+///                         <Comp />
+///                     </PhantomComponent<Provider4>>
```

Comment:
What if `Provider4` is nested deep into the component tree? Do we use conditional compilation or is there a way to detect if the app is being rendered on the server, such as using a context? I assume the context has to be set manually.

Reply:
> I assume the context has to be set manually.

Yes, currently you can manually pass a context.

I will add a hook to read which renderer it is currently using in the future.